OSDOCS-986 GCP UPI shared VPC

kalexand-rh opened the pull request:
See https://issues.redhat.com/browse/OSDOCS-986

kalexand-rh commented:
@jstuever, will you PTAL?

kalexand-rh commented:
@shellyyang1989, will you PTAL?

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+[id="installing-gcp-user-infra-vpc"]
+= Installing a cluster on GCP using Deployment Manager templates
```
Shared VPC is missing in the title. Suggest to update it to Installing a cluster in to a shared VPC on GCP using Deployment Manager templates.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+toc::[]
+
+In {product-title} version {product-version}, you can install a cluster on
+Google Cloud Platform (GCP) that uses infrastructure that you provide.
```
Again, shared vpc is missing.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+In {product-title} version {product-version}, you can install a cluster on
+Google Cloud Platform (GCP) that uses infrastructure that you provide.
+
+The steps for performing a user-provided infrastructure install are outlined here. Several
```
Again, shared vpc is missing. I suggest to update it to The steps for performing a user-provided infrastructure install into a shared VPC are outlined here

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+* Review details about the
+xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update]
+processes.
+* You must have an existing VPC in GCP that contains separate subnets for the compute and control plane machines that the cluster requires.
```
Although the shared vpc alongside host project should be pre-existing resources, IMO, we still need to provide an end-to-end solution to customers so that they can deploy OCP from scratch. Therefore, I suggest to remove this line and add a section to describe how to configure host project and shared VPC.

kalexand-rh replied:
@shellyyang1989 or @jstuever, do you have more details about the host project and VPC configuration?

shellyyang1989 replied:
To learn more about how to enable shared VPC, please refer to https://cloud.google.com/vpc/docs/provisioning-shared-vpc#setting_up. Once shared VPC is enabled, all the vpc networks in host project can be accessed by service project.
The host project is dedicated to host vpc network and dns which do not have much difference with those in stand-alone project.
The host project should have these:
1. Service account which has these roles enabled:
    Compute Network User
    Compute Security Admin
    Deployment Manager Editor
    DNS Administrator
    Security Admin
    Network Management Admin 
2. Have user managed service account of service project got 'Compute Network User' role of host project
3. Have google managed service account i.e. project-number@cloudservices.gserviceaccount.com of service project got 'Compute Network User' role of host project
4. A public dns zone for external dns resolution
5. VPC network

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+and approving them.
+
+[id="installation-gcp-user-infra-config-project-vpc"]
+== Configuring your GCP project
```
In shared VPC scenario, there are 2 kinds of projects: host project which hosts the shared VPC network and service project which hosts the OCP cluster. I think you'd like to describe the service project in this section, and I suggest to update it to Configuring your GCP service project to make it more clear.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+== Configuring your GCP project
+
+Before you can install {product-title}, you must configure a Google Cloud
+Platform (GCP) project to host it.
```
Again, missing service before project.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+the correct machine issued the request. You must determine and implement a
+method of verifying the validity of the kubelet serving certificate requests
+and approving them.
+
```
As mentioned before, add a section "Configuring your GCP host project" here

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+
+include::modules/installation-gcp-project.adoc[leveloffset=+2]
+include::modules/installation-gcp-enabling-api-services.adoc[leveloffset=+2]
+include::modules/installation-gcp-dns.adoc[leveloffset=+2]
```
In shared VPC scenario, host project hosts the DNS zone for the cluster which means DNS is in the different project with OCP cluster. Please update the content in this section accordingly.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+include::modules/installation-gcp-enabling-api-services.adoc[leveloffset=+2]
+include::modules/installation-gcp-dns.adoc[leveloffset=+2]
+include::modules/installation-gcp-limits.adoc[leveloffset=+2]
+include::modules/installation-gcp-service-account.adoc[leveloffset=+2]
```
A service account in host project is also required. And required roles are: 
 Compute Network User
Compute Security Admin
Deployment Manager Editor
DNS Administrator
Security Admin
Network Management Admin

shellyyang1989 commented on modules/installation-gcp-config-yaml.adoc:
```diff
@@ -17,6 +17,11 @@ ifeval::["{context}" == "installing-gcp-private"]
 :private:
```
network, controlPlaneSubnet and computeSubnet is not required in install-config.yaml. If you specify them, you will get the below error as the network is in host project instead of service project. 
FATAL failed to fetch Master Machines: failed to load asset "Install Config": platform.gcp.network: Invalid value: "aos-qe-network": failed to get network aos-qe-network: googleapi: Error 404: The resource 'projects/openshift-qe/global/networks/aos-qe-network' was not found, notFound

shellyyang1989 commented on modules/installation-gcp-config-yaml.adoc:
```diff
@@ -17,6 +17,11 @@ ifeval::["{context}" == "installing-gcp-private"]
 :private:
 :vpc:
```
> Required. The installation program prompts you for this value.
Please explicitly demonstrate it's the DNS zone resided on host project

shellyyang1989 commented on modules/installation-gcp-config-yaml.adoc:
```diff
@@ -17,6 +17,11 @@ ifeval::["{context}" == "installing-gcp-private"]
 :private:
 :vpc:
 endif::[]
```
The replica of work nodes should be 0

shellyyang1989 replied:
Hi @jstuever , please help confirm: in shared VPC scenario, only support to create worker nodes manually, correct? 
In UPI installation of standalone project, cluster could provision the worker nodes, does it work for shared VPC? Actually I tried that but worker nodes were not created. I am not sure if it's expected.

shellyyang1989 commented on modules/installation-gcp-config-yaml.adoc:
```diff
@@ -17,6 +17,11 @@ ifeval::["{context}" == "installing-gcp-private"]
 :private:
 :vpc:
 endif::[]
+ifeval::["{context}" == "installing-gcp-user-infra-vpc"]
```
Regarding this: 
To use an existing VPC in a cluster that uses infrastructure that you provision, you must set publish to Internal

I suggest to replace "an existing" with "a shared". Shared VPC stands for cross project VPC which is not quite same as existing VPC

shellyyang1989 commented on modules/installation-gcp-config-yaml.adoc:
```diff
@@ -118,6 +123,9 @@ For production {product-title} clusters on which you want to perform installatio
 <10> How to publish the user-facing endpoints of your cluster. Set `publish` to `Internal` to deploy a private cluster, which cannot be accessed from the internet. The default value is `External`.
 endif::private[]
+ifdef::user-infra-vpc[]
+To use an existing VPC in a cluster that uses infrastructure that you provision, you must set `publish` to `Internal`. The installation program will no longer be able to access the public DNS zone for the the base domain in the host project.
```
I suggest to replace "an existing" with "a shared". Shared VPC stands for cross project VPC which is not quite same as existing VPC

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,47 @@ spec:
+. Configure the cloud provider for your VPC.
++
+--
+.. Open the `manifests/cloud-provider-config.yml` file.
```
This file is suffixed 'yaml' not yml.

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,47 @@ spec:
++
+--
+.. Open the `manifests/cloud-provider-config.yml` file.
+.. Add the `network-project-id` parameter and set its value to the name of the shared VPC.
```
The value of `network-project-id` is the id of host project

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,47 @@ spec:
+  subnetwork-name = example-worker-subnet
+----
+endif::user-infra-vpc[]
+
```
For external cluster, we need to change load balance to external before ignition files creation:
Open <installation_directory>/manifests/cluster-ingress-default-ingresscontroller.yaml file
Replace the value of 'scope' parameter with External.
The content of this file is like this:
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  creationTimestamp: null
  name: default
  namespace: openshift-ingress-operator
spec:
  endpointPublishingStrategy:
    loadBalancer:
      scope: External
    type: LoadBalancerService
status:
  availableReplicas: 0
  domain: ''
  selector: ''

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,47 @@ spec:
+.. Replace the value of the `subnetwork-name` parameter with the value of the subnet on the shared VPC to host your compute machines.
++
+--
+The contents of the `manifests/cloud-provider-config.yml` resemble the following example:
```
I think absolute path is better.

shellyyang1989 commented:
For $ jq -r .infraID /<installation_directory>/metadata.json 
openshift-vw9j6, remove the first slash would be better.

shellyyang1989 commented:
Regarding 01_VPC.py, it's recommended to use auto_only network. Please get the latest VPC deployment config here: https://github.com/openshift/installer/blob/master/upi/gcp/01_vpc.py.

shellyyang1989 commented:
In Creating networking and load balancing components in GCP section:
again, dns is hosted in host project whereas load balance is hosted in service project. So we need to separate the creation of these 2 resources.
To create private dns zone:
Get 02_dns.py from https://github.com/openshift/installer/blob/master/upi/gcp/02_dns.py
cat <<EOF > ${DIR}/02_dns.yaml
imports:
- path: ${DIR}/02_dns.py
resources:
- name: cluster-dns
  type: ${DIR}/02_dns.py
  properties:
    infra_id: '${INFRA_ID}'
    cluster_domain: '${CLUSTER_NAME}.${BASE_DOMAIN}'
    cluster_network: '${CLUSTER_NETWORK}'
EOF
gcloud deployment-manager deployments create ${INFRA_ID}-dns --config ${DIR}/02_dns.yaml --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}

For external cluster, to create lb:
Get 02_lb_ext.py from https://github.com/openshift/installer/blob/master/upi/gcp/02_lb_ext.py
Get 02_lb_int.py from https://github.com/openshift/installer/blob/master/upi/gcp/02_lb_int.py
cat <<EOF >02_lb.yaml
imports:
- path: 02_lb_ext.py
- path: 02_lb_int.py
resources:
- name: cluster-lb-ext
  type: 02_lb_ext.py
  properties:
    infra_id: '${INFRA_ID}'
    region: '${REGION}'
- name: cluster-lb-int
  type: 02_lb_int.py
  properties:
    cluster_network: '${CLUSTER_NETWORK}'
    control_subnet: '${CONTROL_SUBNET}'
    infra_id: '${INFRA_ID}'
    region: '${REGION}'
    zones:
    - '${ZONE_0}'
    - '${ZONE_1}'
    - '${ZONE_2}'
EOF

# Create the infra resources
gcloud deployment-manager deployments create ${INFRA_ID}-lb --config 02_lb.yaml

export CLUSTER_IP=$(gcloud compute addresses describe ${INFRA_ID}-cluster-ip --region=${REGION} --format json | jq -r .address)
export CLUSTER_PUBLIC_IP=$(gcloud compute addresses describe ${INFRA_ID}-cluster-public-ip --region=${REGION} --format json | jq -r .address)

shellyyang1989 commented on modules/installation-creating-gcp-dns.adoc:
```diff
@@ -33,9 +37,18 @@ requires.
 +
+ifndef::shared-vpc[]
 ----
 $ export CLUSTER_NETWORK=`gcloud compute networks describe ${INFRA_ID}-network --format json | jq -r .selfLink`
```
The network is hosted by host project and current active service account works for service project, so we need to add --project --account to specify host project and service account of host project. For example:
export CLUSTER_NETWORK=$(gcloud compute networks describe ${HOST_PROJECT_NETWORK} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)

shellyyang1989 replied:
@kalexand-rh Please add --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} to the cli:
$ export CLUSTER_NETWORK=`gcloud compute networks describe ${INFRA_ID}-network --format json | jq -r .selfLink`

shellyyang1989 replied:
@kalexand-rh Please be aware of this. This issue has not been addressed yet.

shellyyang1989 commented:
For this step: The templates do not create DNS entries due to limitations of Deployment Manager, so you must create them manually, it should be:
For external cluster, we need to add dns entries to both public zone and private zone:

```javascript
if [ -f transaction.yaml ]; then rm transaction.yaml; fi
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction start --zone ${INFRA_ID}-private-zone
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction add ${CLUSTER_IP} --name api.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction add ${CLUSTER_IP} --name api-int.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction execute --zone ${INFRA_ID}-private-zone

if [ -f transaction.yaml ]; then rm transaction.yaml; fi
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction start --zone ${BASE_DOMAIN_ZONE_NAME}
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction add ${CLUSTER_PUBLIC_IP} --name api.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${BASE_DOMAIN_ZONE_NAME}
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} dns record-sets transaction execute --zone ${BASE_DOMAIN_ZONE_NAME}
```

For internal cluster, only internal dns entries are needed:

```javascript
if [ -f transaction.yaml ]; then rm transaction.yaml; fi
gcloud dns record-sets transaction start --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add ${CLUSTER_IP} --name api.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add ${CLUSTER_IP} --name api-int.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} || exit 2
```

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
 - name: cluster-security
```
The firewall rules are hosted in host project and IAM is hosted in service account. And hence we need to separate their creation. The firewall rules are also different between external cluster and internal cluster. The full steps are like these:
For external cluster, create firewall rules:
Get 03_firewall.py from https://github.com/openshift/installer/blob/master/upi/gcp/03_firewall.py

```javascript
cat <<EOF >${DIR}/03_firewall.yaml
imports:
- path: ${DIR}/03_firewall.py
resources:
- name: cluster-firewall
  type: ${DIR}/03_firewall.py
  properties:
    allowed_external_cidr: '0.0.0.0/0'
    infra_id: '${INFRA_ID}'
    cluster_network: '${CLUSTER_NETWORK}'
    network_cidr: '${NETWORK_CIDR}'
EOF
gcloud deployment-manager deployments create ${INFRA_ID}-firewall --config ${DIR}/03_firewall.yaml --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
```
For internal cluster, create firewall rules:

```javascript
cat <<EOF >${DIR}/03_firewall.yaml
imports:
- path: ${DIR}/03_firewall.py
resources:
- name: cluster-firewall
  type: ${DIR}/03_firewall.py
  properties:
    allowed_external_cidr: '${NETWORK_CIDR}'
    infra_id: '${INFRA_ID}'
    cluster_network: '${CLUSTER_NETWORK}'
    network_cidr: '${NETWORK_CIDR}'
EOF
gcloud deployment-manager deployments create ${INFRA_ID}-firewall --config ${DIR}/03_firewall.yaml --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
```

To create IAM, get 03_iam.py from https://github.com/openshift/installer/blob/master/upi/gcp/03_iam.py

```javascript
cat <<EOF >${DIR}/03_iam.yaml
imports:
- path: ${DIR}/03_iam.py
resources:
- name: cluster-iam
  type: ${DIR}/03_iam.py
  properties:
    infra_id: '${INFRA_ID}'
EOF
gcloud deployment-manager deployments create ${INFRA_ID}-iam --config ${DIR}/03_iam.yaml
```

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+. Export the variable for the master service account:
++
+----
+$ export MASTER_SA=${INFRA_ID}-m@${PROJECT_NAME}.iam.gserviceaccount.com
```
Look up the IAM is better than hard code the dns

    export MASTER_SA=$(gcloud iam service-accounts list --filter "email~^${INFRA_ID}-m@${PROJECT_NAME}." --format json | jq -r '.[0].email')

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+. Export the variable for the master service account:
++
+----
+$ export WORKER_SA=${INFRA_ID}-w@${PROJECT_NAME}.iam.gserviceaccount.com
```
export WORKER_SA=$(gcloud iam service-accounts list --filter "email~^${INFRA_ID}-w@${PROJECT_NAME}." --format json | jq -r '.[0].email')

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+.. Export the variable for the name of the project that hosts the shared VPC:
++
+----
+$ export HOST_PROJECT=<host_project>
```
For HOST_PROJECT_ACCOUNT and HOST_PROJECT, we need them in the previous sections. Please move them forward to Exporting common variables section.

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+----
+$ export HOST_PROJECT=<host_project>
+----
+
```
Master service account need to have below roles of service project:
compute.instanceAdmin, compute.networkAdmin, compute.securityAdmin, iam.serviceAccountUser, storage.admin
```javascript
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${MASTER_SERVICE_ACCOUNT}" --role "roles/compute.instanceAdmin"
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${MASTER_SERVICE_ACCOUNT}" --role "roles/compute.networkAdmin"
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${MASTER_SERVICE_ACCOUNT}" --role "roles/compute.securityAdmin"
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${MASTER_SERVICE_ACCOUNT}" --role "roles/iam.serviceAccountUser"
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${MASTER_SERVICE_ACCOUNT}" --role "roles/storage.admin"
```

Worker service account need to have below roles of service project:
compute.viewer, storage.admin
```javascript
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${WORKER_SERVICE_ACCOUNT}" --role "roles/compute.viewer"
gcloud projects add-iam-policy-binding ${PROJECT_NAME} --member "serviceAccount:${WORKER_SERVICE_ACCOUNT}" --role "roles/storage.admin"
```

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+$ export HOST_PROJECT=<host_project>
+----
+
+.. Grant the `networkViewer` role to the master service account:
```
Grant the `networkViewer` role of host project to the master service account

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+.. Grant the `networkUser` role to the master service account for the control plane subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser"
```
gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser" --region ${REGION}

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+.. Grant the `networkUser` role to the worker service account for the control plane subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser"
```
Need to specify region like this:

    gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser" --region ${REGION}

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+.. Grant the `networkUser` role to the master service account for the compute subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser"
```
Need to specify region like this:

    gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser" --region ${REGION}

shellyyang1989 commented on modules/installation-creating-gcp-security.adoc:
```diff
@@ -48,40 +58,97 @@ resources:
+.. Grant the `networkUser` role to the worker service account for the compute subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser"
```
Need to specify region like this:

    gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser" --region ${REGION}

shellyyang1989 commented on modules/installation-creating-gcp-bootstrap.adoc:
```diff
@@ -32,13 +36,38 @@ have to contact Red Hat support with your installation logs.
 +
+ifndef::shared-vpc[]
 ----
 $ export CONTROL_SUBNET=`gcloud compute networks subnets describe ${INFRA_ID}-master-subnet --region=${REGION} --format json | jq -r .selfLink`
```
Need to specify --project --account:

    export CONTROL_SUBNET=$(gcloud compute networks subnets describe ${HOST_PROJECT_CONTROL_SUBNET} --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)

shellyyang1989 commented on modules/installation-creating-gcp-bootstrap.adoc:
```diff
@@ -105,3 +134,7 @@ $ gcloud compute target-pools add-instances \
 $ gcloud compute target-pools add-instances \
     ${INFRA_ID}-ign-target-pool --instances-zone="${ZONE_0}" --instances=${INFRA_ID}-bootstrap
 ----
+
```
For Add the bootstrap instance to the load balancers step:
1. Add bootstrap instance to internal load balancer instance group:

    gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-bootstrap-instance-group --zone=${ZONE_0} --instances=${INFRA_ID}-bootstrap

2. Add bootstrap instance group to the internal load balancer backend service

    gcloud compute backend-services add-backend ${INFRA_ID}-api-internal-backend-service --region=${REGION} --instance-group=${INFRA_ID}-bootstrap-instance-group --instance-group-zone=${ZONE_0}

shellyyang1989 commented on modules/installation-creating-gcp-bootstrap.adoc:
```diff
@@ -32,13 +36,38 @@ have to contact Red Hat support with your installation logs.
 $ export ZONE_1=`gcloud compute regions describe ${REGION} --format=json | jq -r .zones[1] | cut -d "/" -f9`
+----
++
+----
```
The bootstrap deployment template is changed. please load the latest from https://github.com/openshift/installer/blob/master/upi/gcp/04_bootstrap.py

shellyyang1989 commented:
In the "Creating the control plane machines in GCP" section, it has:
export MASTER_SERVICE_ACCOUNT_EMAIL=`gcloud iam service-accounts list | grep "^${INFRA_ID}-master-node " | awk '{print $2}'`
It seems duplicate with MASTER_SA.

shellyyang1989 commented:
In the "Creating the control plane machines in GCP" section, it has:The templates do not manage load balancer membership due to limitations of Deployment Manager, so you must add the control plane machines manually:
$ gcloud compute target-pools ...
Actually it differs between internal cluster and external cluster.
For external cluster:
```javascript
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_0}-instance-group --zone=${ZONE_0} --instances=${INFRA_ID}-m-0
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_1}-instance-group --zone=${ZONE_1} --instances=${INFRA_ID}-m-1
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_2}-instance-group --zone=${ZONE_2} --instances=${INFRA_ID}-m-2

gcloud compute target-pools add-instances ${INFRA_ID}-api-target-pool --instances-zone="${ZONE_0}" --instances=${INFRA_ID}-m-0
gcloud compute target-pools add-instances ${INFRA_ID}-api-target-pool --instances-zone="${ZONE_1}" --instances=${INFRA_ID}-m-1
gcloud compute target-pools add-instances ${INFRA_ID}-api-target-pool --instances-zone="${ZONE_2}" --instances=${INFRA_ID}-m-2
```
For internal cluster:
```javascript
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_0}-instance-group --zone=${ZONE_0} --instances=${INFRA_ID}-m-0
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_1}-instance-group --zone=${ZONE_1} --instances=${INFRA_ID}-m-1
gcloud compute instance-groups unmanaged add-instances ${INFRA_ID}-master-${ZONE_2}-instance-group --zone=${ZONE_2} --instances=${INFRA_ID}-m-2
```

shellyyang1989 commented:
In the "Wait for bootstrap completion and remove bootstrap resources in GCP" section, it has:
Delete the bootstrap resources:
gcloud compute target-pools ....

It should be:
```javascript
gcloud compute backend-services remove-backend ${INFRA_ID}-api-internal-backend-service --region=${REGION} --instance-group=${INFRA_ID}-bootstrap-instance-group --instance-group-zone=${ZONE_0}
gsutil rm gs://${INFRA_ID}-bootstrap-ignition/bootstrap.ign
gsutil rb gs://${INFRA_ID}-bootstrap-ignition
gcloud deployment-manager deployments delete -q ${INFRA_ID}-bootstrap
```

shellyyang1989 commented on modules/installation-creating-gcp-worker.adoc:
```diff
@@ -39,11 +43,31 @@ have to contact Red Hat support with your installation logs.
 +
+ifndef::shared-vpc[]
 ----
 $ export COMPUTE_SUBNET=`gcloud compute networks subnets describe ${INFRA_ID}-worker-subnet --region=${REGION} --format json | jq -r .selfLink`
```
Need to specify --project --account:
export COMPUTE_SUBNET=$(gcloud compute networks subnets describe <worker-subnet> --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)

shellyyang1989 replied:
I'm sorry that I forgot an option after describe. The correct command line is like this:
export COMPUTE_SUBNET=$(gcloud compute networks subnets describe **${HOST_PROJECT_COMPUTE_SUBNET}** --region=${R    EGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)

kalexand-rh replied:
Because of the if statement format, this change should show up on line 56.

shellyyang1989 commented on modules/installation-creating-gcp-worker.adoc:
```diff
@@ -39,11 +43,31 @@ have to contact Red Hat support with your installation logs.
+.. Export the email address for your service account:
++
+----
 $ export WORKER_SERVICE_ACCOUNT_EMAIL=`gcloud iam service-accounts list | grep "^${INFRA_ID}-worker-node " | awk '{print $2}'`
```
It seems duplicate with WORKER_SA

kalexand-rh replied:
This was an existing command and might be a bug. Do you know which command should be used to generate the worker service email address?

shellyyang1989 replied:
Yeah, I've logged a bz Bug 1852450 - GCP UPI WORKER_SERVICE_ACCOUNT_EMAIL is duplicate with WORKER_SA and MASTER_SERVICE_ACCOUNT_EMAIL is duplicate with MASTER_SA. It's okay to not fix at this point.

shellyyang1989 commented:
For Optional: Adding the ingress DNS records section, in shared VPC scenario, adding the ingress DNS records is a mandatory step since installer doesn't have 'create' permission to dns zone in host project. And internal cluster only requires internal dns records whereas external cluster requires both internal and external dns records.

shellyyang1989 commented:
Before waiting for cluster completion, we need to manually add ingress firewall rules.
For external cluster:

gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute firewall-rules create --allow='tc    p:30000-32767,udp:30000-32767' --network="${CLUSTER_NETWORK}" --source-ranges='130.211.0.0/22,35.191.0.0/16,    209.85.152.0/22,209.85.204.0/22' --target-tags="${INFRA_ID}-master,${INFRA_ID}-worker" ${INFRA_ID}-ingress-h    c
417 
418 gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute firewall-rules create --allow='tc    p:80,tcp:443' --network="${CLUSTER_NETWORK}" --source-ranges="0.0.0.0/0" --target-tags="${INFRA_ID}-master,$    {INFRA_ID}-worker" ${INFRA_ID}-ingress

For internal cluster:
gcloud compute firewall-rules create --allow='tcp:30000-32767,udp:30000-32767' --network="${CLUSTER_NETWORK    }" --source-ranges='130.211.0.0/22,35.191.0.0/16,209.85.152.0/22,209.85.204.0/22' --target-tags="${INFRA_ID}    -master,${INFRA_ID}-worker" ${INFRA_ID}-ingress-hc
383 
gcloud compute firewall-rules create --allow='tcp:80,tcp:443' --network="${CLUSTER_NETWORK}" --source-ranges="${NETWORK_CIDR}" --target-tags="${INFRA_ID}-master,${INFRA_ID}-worker" ${INFRA_ID}-ingress

kalexand-rh commented:
@shellyyang1989, PTAL? I think I've addressed all your comments.

shellyyang1989 commented:
@kalexand-rh Is there an available preview?

kalexand-rh commented:
> @kalexand-rh Is there an available preview?

@shellyyang1989, here is a preview: http://file.rdu.redhat.com/kalexand/070620/osdocs986/installing/installing_gcp/installing-gcp-user-infra-vpc.html

shellyyang1989 commented:
I suggest to update the title to: Installing a cluster into shared VPC on user-provisioned infrastructure in GCP by using Deployment Manager templates

shellyyang1989 commented:
In Configuring the GCP project that hosts your shared VPC section, it says: Create the shared VPC. See Setting up Shared VPC in the GCP documentation. How about changing it to Set up the shared VPC ...?

shellyyang1989 commented:
Again, in Configuring the GCP project that hosts your shared VPC section, it says: The service account for the shared VPC requires the following roles: * Compute Network User. A service account belongs to a project and shared VPC is a kind of concepts. How about changing it to The service account for the project that hosts shared VPC network requires ..., or The service account for the host project requires...?

shellyyang1989 commented:
How about changing the section title 'Configuring the GCP project that hosts your shared VPC' to 'Configuring the GCP project that hosts your shared VPC network'?

shellyyang1989 commented:
With regard to this procedure: Create the shared VPC. See Setting up Shared VPC in the GCP documentation, it needs to enable shared VPC in host project and attach service project to it. So it's better have this procedure behind the configuration of service project.

shellyyang1989 commented:
DNS Administrator is not required by service project and Deployment Manager Editor is required

shellyyang1989 commented:
In "Manually creating the installation configuration file" section, it says: 
For installations of OpenShift Container Platform that use user-provisioned infrastructure, you must manually generate your installation configuration file.
Actually only installation with shared VPC need to Manually generate installation configuration file since installer could not access base domain in host project. So I suggest to add shared VPC in that line.

shellyyang1989 commented:
With regard to "Optional: If you do not want the cluster to provision compute machines, remove the Kubernetes manifest files that define the worker machines:", it's a required procedure in shared VPC since machine controller is not able to automatically provision workers.

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,69 @@ spec:
++
+--
+.. Open the `<installation_directory>/manifests/cloud-provider-config.yaml` file.
+.. Add the `network-project-id` parameter and set its value to the name of project that hoststhe shared VPC.
```
Missing a space: hoststhe

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,69 @@ spec:
+--
+.. Open the `<installation_directory>/manifests/cloud-provider-config.yaml` file.
+.. Add the `network-project-id` parameter and set its value to the name of project that hoststhe shared VPC.
+.. Add the `network-name` parameter and set its value to the name of the network that hosts the shared VPC.
```
The network should host the OCP cluster instead of shared VPC.

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,69 @@ spec:
+.. Open the `<installation_directory>/manifests/cloud-provider-config.yaml` file.
+.. Add the `network-project-id` parameter and set its value to the name of project that hoststhe shared VPC.
+.. Add the `network-name` parameter and set its value to the name of the network that hosts the shared VPC.
+.. Replace the value of the `subnetwork-name` parameter with the value of the subnet on the shared VPC to host your compute machines.
```
I suggest to change it to:
Replace the value of the subnetwork-name parameter with the value of the shared VPC subnet that hosts your compute machines.

shellyyang1989 commented on modules/installation-user-infra-generate-k8s-manifest-ignition.adoc:
```diff
@@ -160,15 +170,69 @@ spec:
+  subnetwork-name = example-worker-subnet
+----
+
+. Open the `<installation_directory>/manifests/cluster-ingress-default-ingresscontroller.yaml` file and replace the value of the `scope` parameter with `External`. The contents of the file resemble the following example:
```
This procedure is only required by external cluster. Private cluster should ignore it.

kalexand-rh commented:
@shellyyang1989, will you take another look? I updated the preview: http://file.rdu.redhat.com/kalexand/070620/osdocs986/installing/installing_gcp/installing-gcp-user-infra-vpc.html

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,138 @@
+[id="installing-gcp-user-infra-vpc"]
+= Installing a cluster with shared a VPC on user-provisioned infrastructure in GCP by using Deployment Manager templates
```
WRT shared a VPC, 'a' is redundant.

shellyyang1989 commented on modules/installation-gcp-user-infra-config-host-project-vpc.adoc:
```diff
@@ -0,0 +1,41 @@
+
+[NOTE]
+====
+If you already have a shared VPC, review this section to ensure that the project that hosts your shared VPC meets all of the requirements to install a {product-title} cluster.
```
If I understand correctly, you want to talk about the existing host project. Then I suggest to change it to:
If you already have a project that hosts shared VPC network, review this section to ensure that the project meets ...

shellyyang1989 commented on modules/installation-gcp-user-infra-config-host-project-vpc.adoc:
```diff
@@ -0,0 +1,41 @@
+* Network Management Admin
+====
+
+. Create the shared VPC. See link:https://cloud.google.com/vpc/docs/provisioning-shared-vpc#setting_up[Setting up Shared VPC] in the GCP documentation.
```
This procedure includes enabling shared VPC in host project and attaching service project to it. So it's better having this procedure behind the configuration of service project that hosts your OCP cluster. And it's better replacing Create with Set up.

shellyyang1989 commented on modules/installation-initializing-manual.adoc:
```diff
@@ -17,7 +17,7 @@ endif::[]
 
 For installations of {product-title} that use user-provisioned
-infrastructure, you must manually generate your installation configuration file.
+infrastructure, you manually generate your installation configuration file.
```
Since installer is not able to access the dns zone in host project, I suggest to demonstrate that **with shared VPC**, you must manually generate your installation configuration file.

shellyyang1989 commented on modules/installation-user-infra-exporting-common-variables.adoc:
```diff
@@ -54,6 +67,20 @@ $ export REGION=`jq -r .gcp.region <installation_directory>/metadata.json`
 ----
 <1> For `<installation_directory>`, specify the path to the directory that you stored the installation files in.
 
+ifdef::shared-vpc[]
```
Could you please add a note to explain that base_domain is the public dns zone on host project?

shellyyang1989 commented on modules/installation-user-infra-exporting-common-variables.adoc:
```diff
@@ -54,6 +67,20 @@ $ export REGION=`jq -r .gcp.region <installation_directory>/metadata.json`
+$ export HOST_PROJECT_ACCOUNT=<host_project_account>
+----
+
+. Export the variable for the name of the project that hosts the shared VPC:
```
Again, should be the **ID** of the project that hosts the shared VPC network

shellyyang1989 commented on modules/installation-user-infra-exporting-common-variables.adoc:
```diff
@@ -54,6 +67,20 @@ $ export REGION=`jq -r .gcp.region <installation_directory>/metadata.json`
 <1> For `<installation_directory>`, specify the path to the directory that you stored the installation files in.
 
+ifdef::shared-vpc[]
+. Export the variable for the name of the account that hosts the shared VPC:
```
It should be the **email** of the service account that belongs to host project

kalexand-rh replied:
It's weird that this variable doesn't include `_EMAIL` like the other variables that expect an email address.

shellyyang1989 commented on modules/installation-deployment-manager-vpc.adoc:
```diff
@@ -67,9 +54,8 @@ def GenerateConfig(context):
+                'minPortsPerVm': 512,
                 'sourceSubnetworkIpRangesToNat': 'LIST_OF_SUBNETWORKS',
                 'subnetworks': [{
                     'name': '$(ref.' + context.properties['infra_id'] + '-worker-subnet.selfLink)',
```
If we cover the VPC network creation, that would help build a better user experience. So could you please add a section to introduce the way to create VPC network in host project? The procedure is like this:
```javascript
$ cat <<EOF >01_vpc.yaml
imports:
- path: 01_vpc.py

resources:
- name: cluster-vpc
  type: 01_vpc.py
  properties:
    infra_id: '${INFRA_ID}' 
    region: '${REGION}' 
    master_subnet_cidr: '${MASTER_SUBNET_CIDR}' 
    worker_subnet_cidr: '${WORKER_SUBNET_CIDR}' 
EOF

$ gcloud deployment-manager deployments create ${INFRA_ID}-vpc --config 01_vpc.yaml --project ${HOST_PROJECT}     --account ${HOST_PROJECT_ACCOUNT}
```

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,118 @@
+
+. For an external cluster, copy the template from the *Deployment Manager template for the external load balancer*
+section of this topic and save it as `02_lb_ext.py` on your computer. This
+template describes the internal load balancing objects that your cluster
```
02_lb_ext.py describes the external load balance.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,118 @@
+  type: 02_lb_int.py
+  properties:
+    cluster_network: '${CLUSTER_NETWORK}' <1>
+    control_subnet: '${CONTROL_SUBNET}' <2>
```
CONTROL_SUBNET is not defined yet.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,118 @@
+    zones:
+    - '${ZONE_0}'
+    - '${ZONE_1}'
+    - '${ZONE_2}'
```
CONTROL_SUBNET, ZONE_0, ZONE_1, ZONE_1 are not defined yet

shellyyang1989 commented on modules/installation-deployment-manager-ext-lb.adoc:
```diff
@@ -0,0 +1,49 @@
+            'region': context.properties['region']
+        }
+    }, {
+        # Refer to docs/dev/kube-apiserver-health-check.md on how to correctly setup health check probe for kube-apiserver
```
It's better to remove this comment.

shellyyang1989 commented on modules/installation-deployment-manager-int-lb.adoc:
```diff
@@ -0,0 +1,84 @@
+            'subnetwork': context.properties['control_subnet']
+        }
+    }, {
+        # Refer to docs/dev/kube-apiserver-health-check.md on how to correctly setup health check probe for kube-apiserver
```
It's better to remove this comment.

shellyyang1989 commented on modules/installation-creating-gcp-private-dns.adoc:
```diff
@@ -0,0 +1,91 @@
+.Procedure
+
+. Copy the template from the *Deployment Manager template for the private DNS*
+section of this topic and save it as `02_infra.py` on your computer. This
```
The file is named 02_infra.py but 02_dns.py is used in procedure 3. Please make them consistent.

shellyyang1989 commented on modules/installation-creating-gcp-private-dns.adoc:
```diff
@@ -0,0 +1,91 @@
+
+. Copy the template from the *Deployment Manager template for the private DNS*
+section of this topic and save it as `02_infra.py` on your computer. This
+template describes the networking and load balancing objects that your cluster
```
The template only describes the private dns zone objects.

shellyyang1989 commented on modules/installation-creating-gcp-private-dns.adoc:
```diff
@@ -0,0 +1,91 @@
+. Export the following variable required by the resource definition:
++
+----
+$ export CLUSTER_NETWORK=`gcloud compute networks describe ${INFRA_ID}-network --format json | jq -r .selfLink`
```
It's already defined in Creating load balancers in GCP section. So please remove it.

shellyyang1989 commented on modules/installation-creating-gcp-private-dns.adoc:
```diff
@@ -0,0 +1,91 @@
+.. Export the following variable:
++
+----
+$ export CLUSTER_IP=`gcloud compute addresses describe ${INFRA_ID}-cluster-public-ip --region=${REGION} --format json | jq -r .address`
```
It's already defined in Creating load balancers in GCP section. So please remove it.

shellyyang1989 commented on modules/installation-creating-gcp-private-dns.adoc:
```diff
@@ -0,0 +1,91 @@
+$ gcloud dns record-sets transaction start --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
+$ gcloud dns record-sets transaction add ${CLUSTER_IP} --name api.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
+$ gcloud dns record-sets transaction add ${CLUSTER_IP} --name api-int.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
+$ gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} || exit 2
```
Please remove "|| exit 2".

shellyyang1989 commented on modules/installation-creating-gcp-firewall-rules-vpc.adoc:
```diff
@@ -0,0 +1,66 @@
+. Copy the template from the
+*Deployment Manager template for firewall rules*
+section of this topic and save it as `03_firewall.py` on your computer. This
+template describes the security groups and roles that your cluster requires.
```
This template only describes the security groups.

shellyyang1989 commented on modules/installation-creating-gcp-firewall-rules-vpc.adoc:
```diff
@@ -0,0 +1,66 @@
++
+----
+$ export MASTER_NAT_IP=`gcloud compute addresses describe ${INFRA_ID}-master-nat-ip --region ${REGION} --format json | jq -r .address`
+$ export WORKER_NAT_IP=`gcloud compute addresses describe ${INFRA_ID}-worker-nat-ip --region ${REGION} --format json | jq -r .address`
```
MASTER_NAT_IP, WORKER_NAT_IP are not used in the whole paper.

shellyyang1989 commented on modules/installation-creating-gcp-firewall-rules-vpc.adoc:
```diff
@@ -0,0 +1,66 @@
+- name: cluster-firewall
+  type: 03_firewall.py
+  properties:
+    allowed_external_cidr: '0.0.0.0/0' <1>
```
The value of allowed_external_cidr differs between external cluster and internal cluster. For external cluster, it's '0.0.0.0/0'. For internal cluster, it's '${NETWORK_CIDR}'

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
+. Copy the template from the
+*Deployment Manager template for IAM roles*
+section of this topic and save it as `03_iam.py` on your computer. This
+template describes the security groups and roles that your cluster requires.
```
This template only describes the roles.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
++
+----
+$ export MASTER_NAT_IP=`gcloud compute addresses describe ${INFRA_ID}-master-nat-ip --region ${REGION} --format json | jq -r .address`
+$ export WORKER_NAT_IP=`gcloud compute addresses describe ${INFRA_ID}-worker-nat-ip --region ${REGION} --format json | jq -r .address`
```
These 2 variables are not used in the whole paper.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
+.. Grant the `networkUser` role to the master service account for the control plane subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser" --region ${REGION}
```
CONTROL_SUBNET is defined instead of HOST_PROJECT_CONTROL_SUBNET.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
+.. Grant the `networkUser` role to the worker service account for the control plane subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_CONTROL_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser" --region ${REGION}
```
Again, CONTROL_SUBNET is defined instead of HOST_PROJECT_CONTROL_SUBNET.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
+.. Grant the `networkUser` role to the master service account for the compute subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${MASTER_SA}" --role "roles/compute.networkUser" --region ${REGION}
```
Again, CONTROL_SUBNET is defined instead of HOST_PROJECT_CONTROL_SUBNET.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,123 @@
+.. Grant the `networkUser` role to the worker service account for the compute subnet:
++
+----
+$ gcloud --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} compute networks subnets add-iam-policy-binding "${HOST_PROJECT_COMPUTE_SUBNET}" --member "serviceAccount:${WORKER_SA}" --role "roles/compute.networkUser" --region ${REGION}
```
HOST_PROJECT_COMPUTE_SUBNET has not been defined yet. Please keep the name convention consistent with CONTROL_SUBNET.

shellyyang1989 commented on modules/installation-creating-gcp-bootstrap.adoc:
```diff
@@ -105,3 +134,21 @@ $ gcloud compute target-pools add-instances \
 $ gcloud compute target-pools add-instances \
     ${INFRA_ID}-ign-target-pool --instances-zone="${ZONE_0}" --instances=${INFRA_ID}-bootstrap
 ----
+
```
Adding bootstrap to external load balancer target pool is not required. So please remove step 7:


The templates do not manage load balancer membership due to limitations of Deployment Manager, so you must add the bootstrap machine manually:

$ gcloud compute target-pools add-instances \
    ${INFRA_ID}-api-target-pool --instances-zone="${ZONE_0}" --instances=${INFRA_ID}-bootstrap
$ gcloud compute target-pools add-instances \
    ${INFRA_ID}-ign-target-pool --instances-zone="${ZONE_0}" --instances=${INFRA_ID}-bootstrap

shellyyang1989 commented on modules/installation-creating-gcp-control-plane.adoc:
```diff
@@ -102,6 +107,7 @@ $ gcloud dns record-sets transaction add \
 $ gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone
```
As the private zone is in host project, need to add --project --account to these cli.
```javascript
if [ -f transaction.yaml ]; then rm transaction.yaml; fi
gcloud dns record-sets transaction start --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add ${MASTER0_IP} --name etcd-0.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add ${MASTER1_IP} --name etcd-1.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add ${MASTER2_IP} --name etcd-2.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction add \
  "0 10 2380 etcd-0.${CLUSTER_NAME}.${BASE_DOMAIN}." \
  "0 10 2380 etcd-1.${CLUSTER_NAME}.${BASE_DOMAIN}." \
  "0 10 2380 etcd-2.${CLUSTER_NAME}.${BASE_DOMAIN}." \
  --name _etcd-server-ssl._tcp.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 60 --type SRV --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
```

shellyyang1989 commented on modules/installation-gcp-user-infra-adding-ingress.adoc:
```diff
@@ -62,3 +72,7 @@ alertmanager-main-openshift-monitoring.apps.your.cluster.domain.example.com
 grafana-openshift-monitoring.apps.your.cluster.domain.example.com
 prometheus-k8s-openshift-monitoring.apps.your.cluster.domain.example.com
 ----
+
```
Public dns entries are only required by external cluster. Internal cluster only requires internal dns entries.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,138 @@
+If you chose to create each rule based on events, you must create firewall rules after you provision the cluster and during the life of the cluster when the console notifies you that rules are missing. Events that are similar to the following event are displayed, and you must add the firewall rules that are required:
+
+----
+Firewall change required by security admin: `gcloud compute firewall-rules create k8s-fw-a26e631036a3f46cba28f8df67266d55 --network example-network --description "{\"kubernetes.io/service-name\":\"openshift-ingress/router-default\", \"kubernetes.io/service-ip\":\"35.237.236.234\"}\" --allow tcp:443,tcp:80 --source-ranges 0.0.0.0/0 --target-tags exampl-fqzq7-master,exampl-fqzq7-worker --project ex ...
```
The cluster has trouble emitting events currently. We have a bug to track it. I'm not sure if it's proper to have a problematic function described here.
Bug 1816806 - system:serviceaccount:kube-system:cloud-provider cannot create resource events

kalexand-rh replied:
[This comment](https://bugzilla.redhat.com/show_bug.cgi?id=1816806#c12) on the bug makes it sound like it's fixed in 4.5. If it's still an issue, I will open a separate PR to fix it.

shellyyang1989 commented on modules/installation-creating-gcp-shared-vpc-cluster-wide-firewall-rules.adoc:
```diff
@@ -0,0 +1,30 @@
+. Add a single firewall rule to allow the Google Cloud Engine health checks to access all of the services. This rule enables the ingress load balancers to determine the health status of their instances.
++
+----
+$ gcloud compute firewall-rules create --allow='tcp:30000-32767,udp:30000-32767' --network="${CLUSTER_NETWORK}" --source-ranges='130.211.0.0/22,35.191.0.0/16,209.85.152.0/22,209.85.204.0/22' --target-tags="${INFRA_ID}-master,${INFRA_ID}-worker" ${INFRA_ID}-ingress-hc
```
Please add --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} to this command line.

shellyyang1989 commented on modules/installation-creating-gcp-shared-vpc-cluster-wide-firewall-rules.adoc:
```diff
@@ -0,0 +1,30 @@
+. Add a single firewall rule to allow access to all cluster services:
++
+----
+$ gcloud compute firewall-rules create --allow='tcp:80,tcp:443' --network="${CLUSTER_NETWORK}" --source-ranges="0.0.0.0/0" --target-tags="${INFRA_ID}-master,${INFRA_ID}-worker" ${INFRA_ID}-ingress
```
Again, please add --account=${HOST_PROJECT_ACCOUNT} --project=${HOST_PROJECT} to this command line.

shellyyang1989 commented on modules/installation-creating-gcp-shared-vpc-cluster-wide-firewall-rules.adoc:
```diff
@@ -0,0 +1,30 @@
+$ gcloud compute firewall-rules create --allow='tcp:80,tcp:443' --network="${CLUSTER_NETWORK}" --source-ranges="0.0.0.0/0" --target-tags="${INFRA_ID}-master,${INFRA_ID}-worker" ${INFRA_ID}-ingress
+----
++
+If you want your cluster to be private, you can add the `--source-ranges=${NETWORK_CIDR}` option.
```
Since --source-ranges is an option in the command line, for private cluster, we need to change the value instead of add the option.

shellyyang1989 commented on modules/installation-creating-gcp-shared-vpc-ingress-firewall-rules.adoc:
```diff
@@ -0,0 +1,32 @@
+// Module included in the following assemblies:
```
This section is duplicate with Creating cluster-wide firewall rules for a shared VPC in GCP section. Please check.

kalexand-rh commented:
@shellyyang1989, I think I've made updates for all of your new comments except for the one about the event bug. Thank you so much. I updated the preview: http://file.rdu.redhat.com/kalexand/070620/osdocs986/installing/installing_gcp/installing-gcp-user-infra-vpc.html

shellyyang1989 commented on modules/installation-gcp-user-infra-shared-vpc-config-yaml.adoc:
```diff
@@ -0,0 +1,76 @@
+sshKey: ssh-ed25519 AAAA... <6>
+publish: Internal <7>
+----
+<1> Specify the public DNS zone on host project.
```
Please remove zone.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,136 @@
+include::modules/installation-user-infra-generate-k8s-manifest-ignition.adoc[leveloffset=+2]
+.Additional resources
+
+* xref:../../installing/installing_gcp/installing-gcp-user-infra.adoc#installation-gcp-user-infra-adding-ingress_installing-gcp-user-infra[Adding the ingress DNS records]
```
@kalexand-rh There is a section about adding the ingress DNS records in the end. I mean it seems not need to have it linked here. Please check.

kalexand-rh replied:
Ah, sorry for misunderstanding. I'm removing the link and putting the module back.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+.. Export the cluster network location:
++
+----
+$ export CLUSTER_NETWORK=`gcloud compute networks describe ${INFRA_ID}-network --format json | jq -r .selfLink`
```
Please add --project ${HOST_PROJECT}     --account ${HOST_PROJECT_ACCOUNT} to this command line.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+
+.. Export the control plane subnet location:
+----
+$ export CONTROL_SUBNET=$(gcloud compute networks subnets describe <master-subnet> --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)
```
As we create VPC network in the last section, the name of master subnet is ${INFRA_ID}-master-subnet. And to keep it consistent with command line $ export CLUSTER_NETWORK=..., I suggest to change  <master-subnet> to ${INFRA_ID}-master-subnet.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+Where `<master-subnet>` is the name of the subnet in the shared VPC to deploy the control plane machines to.
+
+
+.. Export each zone that the cluster uses:
```
Please keep an eye on the number, as well as the format. It should be c instead of a, right?

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+$ export ZONE_2=`gcloud compute regions describe ${REGION} --format=json | jq -r .zones[2] | cut -d "/" -f9`
+----
+
+. Create a `02_lb.yaml` resource definition file:
```
Again, Please keep an eye on the number, as well as the format.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+<3> `cluster_network` is the `selfLink` URL to the cluster network.
+<4> `control_subnet` is the `selfLink` URL to the control subnet.
+
+. Create the deployment by using the `gcloud` CLI:
```
Again, Please keep an eye on the number, indentation, as well as the format.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+$ gcloud deployment-manager deployments create ${INFRA_ID}-lb --config 02_lb.yaml
+----
+
+. Export the cluster IP address:
```
Again, Please keep an eye on the number, indentation, as well as the format.

kalexand-rh replied:
I think this section is right now.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+$ export CLUSTER_IP=$(gcloud compute addresses describe ${INFRA_ID}-cluster-ip --region=${REGION} --format json | jq -r .address)
+----
+
+. Export the cluster public IP address:
```
Again, Please keep an eye on the number, indentation, as well as the format.

kalexand-rh replied:
I think that the indentation is right now and that the problem is that it needs to say "For an external cluster, export..."

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,128 @@
+.. Export the name of the host project control plane subnet:
++
+----
+$ export HOST_PROJECT_CONTROL_SUBNET=<control_plane_subnet>
```
As the VPC network is created by google deployment template in the previous section, we actually know the name of subnet. So I suggest to change it to:
```javascript
$ export HOST_PROJECT_CONTROL_SUBNET="${INFRA_ID}-master-subnet"
```

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,128 @@
+.. Export the name of the host project control plane subnet:
++
+----
+$ export HOST_PROJECT_COMPUTE_SUBNET=<compute_subnet>
```
$ export HOST_PROJECT_COMPUTE_SUBNET="${INFRA_ID}-worker-subnet"

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,128 @@
+----
+$ export HOST_PROJECT_CONTROL_SUBNET=<control_plane_subnet>
+----
+.. Export the name of the host project control plane subnet:
```
Please change control plan to compute.

shellyyang1989 commented on modules/installation-creating-gcp-vpc.adoc:
```diff
@@ -2,6 +2,11 @@
+endif::[]
 
 [id="installation-creating-gcp-vpc_{context}"]
 = Creating a VPC in GCP
```
We use the network when generating ignition files. So we need to move it ahead of that section. I suggest to have it as a 2-level section under Configuring the GCP project that hosts your shared VPC network. When moving the section, please keep an eye on the variables in 01_vpc.yaml to make sure they're defined before or we can use <> to describe them directly.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+.. Export the cluster network location:
++
+----
+$ export CLUSTER_NETWORK=`gcloud compute networks describe ${INFRA_ID}-network --format json | jq -r .selfLink`
```
--project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} is required. If network variables are defined in Creating VPC network section, we can invoke it here like this:
```javascript
export CLUSTER_NETWORK=`gcloud compute networks describe ${HOST_PROJECT_NETWORK} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
```

shellyyang1989 replied:
@kalexand-rh Please be aware of this ^^. It has not been updated yet.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,141 @@
+
+.. Export the control plane subnet location:
+----
+$ export CONTROL_SUBNET=$(gcloud compute networks subnets describe <master-subnet> --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)
```
For the syntax both $() and are correct. In this paper some command lines are using $() and the others are using ``. It's better to keep them consistent.
Please also keep an eye on the format, it seems space or tab is missing in the front of this line.

If network variables are defined in Creating VPC network section, we can invoke it here like this:
```javascript
export CONTROL_SUBNET=`gcloud compute networks subnets describe ${HOST_PROJECT_CONTROL_SUBNET} --region=${REGION} --project ${HOST_ PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
```

kalexand-rh replied:
I'm planning to open a follow-up PR to make updates to these modules so they better align with some of our docs guidelines, and I can take a look the overall command syntax then.

shellyyang1989 commented on modules/installation-creating-gcp-worker.adoc:
```diff
@@ -39,11 +43,31 @@ have to contact Red Hat support with your installation logs.
+endif::shared-vpc[]
+ifdef::shared-vpc[]
+----
+$ export COMPUTE_SUBNET=$(gcloud compute networks subnets describe --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink)
```
Missing subnet behind describe. Again, if you export network variables before, we can invoke it here:
```javascript
$ export COMPUTE_SUBNET=`gcloud compute networks subnets describe ${HOST_PROJECT_COMPUTE_SUBNET} --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
```

shellyyang1989 commented on modules/installation-gcp-user-infra-adding-ingress.adoc:
```diff
@@ -34,23 +44,31 @@ NAME             TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)
 $ if [ -f transaction.yaml ]; then rm transaction.yaml; fi
 $ gcloud dns record-sets transaction start --zone ${INFRA_ID}-private-zone
 $ gcloud dns record-sets transaction add ${ROUTER_IP} --name \*.apps.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 300 --type A --zone ${INFRA_ID}-private-zone
 $ gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone
```
Missing --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} to these command lines.

shellyyang1989 commented on modules/installation-gcp-user-infra-adding-ingress.adoc:
```diff
@@ -34,23 +44,31 @@ NAME             TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)
+$ if [ -f transaction.yaml ]; then rm transaction.yaml; fi
+$ gcloud dns record-sets transaction start --zone ${BASE_DOMAIN_ZONE_NAME}
+$ gcloud dns record-sets transaction add ${ROUTER_IP} --name \*.apps.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 300 --type A --zone ${BASE_DOMAIN_ZONE_NAME}
+$ gcloud dns record-sets transaction execute --zone ${BASE_DOMAIN_ZONE_NAME}
```
Missing --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} to these command lines.

kalexand-rh commented:
@shellyyang1989, I think I've addressed all of your new comments except for the one about standardizing the command syntax, which I think would be better to address on a separate issue. Thank you so much for being so thorough with your checks and patient with your updates. I updated the preview: http://file.rdu.redhat.com/kalexand/070620/osdocs986/installing/installing_gcp/installing-gcp-user-infra-vpc.html

Will you please let me know what else you see?

shellyyang1989 commented on modules/installation-creating-gcp-vpc.adoc:
```diff
@@ -40,19 +69,58 @@ resources:
     infra_id: '${INFRA_ID}' <1>
+endif::shared-vpc[]
+ifdef::shared-vpc[]
+    infra_id: '<prefix> '<1>
```
There is a redundant space between > and '.

shellyyang1989 commented on modules/installation-creating-gcp-vpc.adoc:
```diff
@@ -40,19 +69,58 @@ resources:
+endif::shared-vpc[]
+ifdef::shared-vpc[]
+----
+$ gcloud deployment-manager deployments create ${INFRA_ID}-vpc --config 01_vpc.yaml --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
```
Sorry for misleading you. As the variables INFRA_ID is not available here, we can change ${INFRA_ID}-vpc to <vpc_deployment_name> or something else. And we need to define HOST_PROJECT and HOST_PROJECT_ACCOUNT before this command line.

shellyyang1989 commented on modules/installation-creating-gcp-vpc.adoc:
```diff
@@ -40,19 +69,58 @@ resources:
+----
+$ export HOST_PROJECT_CONTROL_SUBNET=<control_plane_subnet>
+----
+.. Export the name of the host project control plane subnet:
```
Please change control plane to compute.

shellyyang1989 commented on modules/installation-gcp-user-infra-shared-vpc-config-yaml.adoc:
```diff
@@ -0,0 +1,76 @@
+platform:
+  gcp:
+    ProjectID: openshift-production
+    region: us-central1
```
As we define REGION in the Creating the VPC section, please add a note below to instruct that the region should be consistent with REGION. I mean the region in which the cluster will install should be the same as the region that VPC network resides in.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,142 @@
+.. Export the control plane subnet location:
++
+----
+$ export CONTROL_SUBNET=`gcloud compute networks subnets describe ${HOST_PROJECT_CONTROL_SUBNET} --region=${REGION} --project ${HOST_ PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
```
There is a redundant space in HOST_ PROJECT

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,142 @@
+$ export CONTROL_SUBNET=`gcloud compute networks subnets describe ${HOST_PROJECT_CONTROL_SUBNET} --region=${REGION} --project ${HOST_ PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
+----
++
+Where `<master-subnet>` is the name of the subnet in the shared VPC to deploy the control plane machines to.
```
This is not required since master-subnet is not used.

shellyyang1989 commented on modules/installation-creating-gcp-iam-shared-vpc.adoc:
```diff
@@ -0,0 +1,116 @@
+// * installing/installing_gcp/installing-gcp-user-infra-vpc.adoc
+
+[id="installation-creating-gcp-iam-shared-vpc_{context}"]
+= Creating IAM roles in GCP
```
As HOST_PROJECT_CONTROL_SUBNET and HOST_PROJECT_COMPUTE_SUBNET are defined in Creating VPC section, please remove them from this section.

kalexand-rh replied:
I only see where these variables are used in this file now, so if you're looking through the preview, it might not have refreshed right. I'm uploading a fresh preview now.

shellyyang1989 commented on modules/installation-creating-gcp-worker.adoc:
```diff
@@ -39,11 +43,31 @@ have to contact Red Hat support with your installation logs.
+$ export COMPUTE_SUBNET=`gcloud compute networks subnets describe ${HOST_PROJECT_COMPUTE_SUBNET} --region=${REGION} --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT} --format json | jq -r .selfLink`
+----
++
+Where `<worker-subnet>` is the name of the subnet in the shared VPC to deploy the compute machines to.
```
worker-subnet is not used so please remove this line.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,142 @@
+template describes the internal load balancing objects that your cluster
+requires.
+
+. For an external cluster, copy the template from the *Deployment Manager template for the external load balancer*
```
Adding also before copy might make it more clear.

shellyyang1989 commented on modules/installation-creating-gcp-lb.adoc:
```diff
@@ -0,0 +1,142 @@
+$ export CLUSTER_IP=$(gcloud compute addresses describe ${INFRA_ID}-cluster-ip --region=${REGION} --format json | jq -r .address)
+----
+
+. For an external cluster, export the cluster public IP address:
```
Adding also before export might make it more clear.

shellyyang1989 commented on modules/installation-gcp-user-infra-adding-ingress.adoc:
```diff
@@ -34,23 +44,31 @@ NAME             TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)
+$ gcloud dns record-sets transaction add ${ROUTER_IP} --name \*.apps.${CLUSTER_NAME}.${BASE_DOMAIN}. --ttl 300 --type A --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
+$ gcloud dns record-sets transaction execute --zone ${INFRA_ID}-private-zone --project ${HOST_PROJECT} --account ${HOST_PROJECT_ACCOUNT}
 ----
+... For an external cluster, add the A record to the public zones:
```
Adding also before add might make it more clear.

shellyyang1989 commented on modules/installation-creating-gcp-shared-vpc-ingress-firewall-rules.adoc:
```diff
@@ -0,0 +1,32 @@
+// * installing/installing_gcp/installing-gcp-user-infra-vpc.adoc
+
+[id="installation-creating-gcp-shared-vpc-ingress-firewall-rules_{context}"]
+= Creating ingress firewall rules for a shared VPC in GCP
```
This section seems duplicate with Creating cluster-wide firewall rules for a shared VPC in GCP. Please check.

kalexand-rh replied:
@shellyyang1989, it looks like the options in the commands are different, and I am not sure if they're trying to do exactly the same thing or not.

I don't work with the devs who work on ingress, so you know more about this than I do. What can I do today to make this work for GA, and will you file a bug for what you think we need to figure out to make this section right?

shellyyang1989 replied:
@kalexand-rh I don't find the command lines are different. If you're seeing any difference, please let me know.

shellyyang1989 commented on installing/installing_gcp/installing-gcp-user-infra-vpc.adoc:
```diff
@@ -0,0 +1,135 @@
+
+include::modules/installation-gcp-user-infra-adding-ingress.adoc[leveloffset=+1]
+
+[id="installation-gcp-user-infra-vpc-adding-firewall-rules"]
```
Per upstream doc, there are 2 approaches to add firewall rules: 1 is Adding firewall rules based on cluster events and another is Adding a cluster-wide firewall rules. So I suggest to change the section name as below:
Adding the ingress firewall rules
    Adding firewall rules based on cluster events
    Adding cluster-wide firewall rules (alternative)

kalexand-rh replied:
I checked with the author of the dev docs, and dev's preference is to use the event-based rules to better mimic the IPI functionality.  I structured the docs to mimic that.

shellyyang1989 replied:
I see. What I mean was that we need to instruct that Adding cluster-wide firewall rules is an alternative of event-based rules, if there is any issue on event, user can add cluster-wide firewall rules instead.

shellyyang1989 commented:
LGTM except for the copy buttons and the zone example for bootstrap, master, worker.

kalexand-rh commented:
Great! I am going to fix the copy buttons on a separate PR because it's going to involve a lot of code block changes. I fixed the zones to use us-central1. Per [the GCP docs](https://cloud.google.com/compute/docs/regions-zones), us-central1 uses a,b,c,f instead of b,c,d, like the us-east1 examples, so I'm adjusting those references, too.

kalexand-rh commented:
@vikram-redhat approved merging this change with the final amendments.

kalexand-rh commented:
/cherrypick enterprise-4.5
